storage_scrubber/src/pageserver_physical_gc.rs

```diff
@@ -597,6 +602,60 @@ async fn gc_timeline(
         }
     };
 
+    if let Some(tenant_manifest_info) = &*tenant_manifest_info {
+        // TODO: this is O(n^2) in the number of offloaded timelines. Do a hashmap lookup instead.
+        let maybe_offloaded = tenant_manifest_info
+            .manifest
+            .offloaded_timelines
+            .iter()
+            .find(|offloaded_timeline| offloaded_timeline.timeline_id == ttid.timeline_id);
+        if let Some(offloaded) = maybe_offloaded {
+            let warnings = validate_index_part_with_offloaded(index_part, offloaded);
+            let warn = if warnings.is_empty() {
+                false
+            } else {
+                // Verify that the manifest hasn't changed. If it has, a potential racing change could have been cause for our troubles.
+                match list_tenant_manifests(remote_client, ttid.tenant_shard_id, target).await? {
+                    ListTenantManifestResult::WithErrors {
+                        errors,
+                        unknown_keys: _,
+                    } => {
+                        for (_key, error) in errors {
+                            tracing::warn!(%ttid, "list_tenant_manifests in gc_timeline: {error}");
+                        }
+                        true
+                    }
+                    ListTenantManifestResult::NoErrors {
+                        latest_generation,
+                        manifests: _,
+                    } => {
+                        if let Some(new_latest_gen) = latest_generation {
+                            let manifest_changed = (
+                                new_latest_gen.generation,
+                                new_latest_gen.listing_object.last_modified,
+                            ) == (
+                                tenant_manifest_info.generation,
+                                tenant_manifest_info.listing_object.last_modified,
+                            );
+                            if manifest_changed {
+                                tracing::debug!(%ttid, "tenant manifest changed since it was loaded, suppressing {} warnings", warnings.len());
+                            }
+                            manifest_changed
```

Comment:
Instead of comparing `generation,last_modified`, why not simply call `validate_index_part_with_offloaded` again?

The scenario I pointed out in the previous review would not result in a second validation failure, I think.

Reply:
Then the `index_part.json` could have had an upload since we've downloaded the manifest. Even if we always download the manifest after index_part.json, the timeline that we have just unarchived, could become archived again, and if we are fast enough, the compaction might still be ongoing so it might actually immediately offload it. 

If the timeline is archived and then immediately offloaded (say compaction runs at a particularly bad time) between the index_part.json download and the manifest download, we'd have another false positive.

The index_part.json would then have it still as unarchived, but the manifest would have it as offloaded. Admittedly it is unlikely as the compaction job usually doesn't run right away.

What we want is an atomic view into both `index_part.json` and the manifest at the same time T, because our invariant is formulated to hold over all possible T, but looks at the relation of files from the same T.

Of course we don't have such a primitive, but we have a "has the file changed since I last looked at it" primitive. With that second primitive, we can ensure that the manifest has not changed in the entire period from A to C, where A and C are timestamps where we downloaded the manifest, so therefore, the manifest must also have had that same state at timepoint B for which A<B<C holds.

We could extend this idea to a loop that downloads first index_part and then the manifest until either of both has been observed for two timestamps where it hasn't changed, and the other file has been observed between.

But personally I don't think it's worth that effort: we run the scrubber regularly so we can look at it the next time we get to it. And it's not guaranteed to terminate as the timeline plus tenant could be "busy".